services/webnn/ort/graph_impl_ort.cc

```diff
@@ -126,8 +131,34 @@ GraphImplOrt::CreateAndBuildOnBackgroundThread(
     // supports it.
   }
 
-  // TODO(https://github.com/shiyi9801/chromium/issues/46): Append OpenVINO EP
-  // for GPU and NPU devices.
+  // Select the execution provider.
+  switch (context_options->device) {
+    case mojom::CreateContextOptions::Device::kCpu: {
+      // TODO: Investigate how to apply layout optimizations (ORT_ENABLE_ALL):
+      // https://onnxruntime.ai/docs/performance/model-optimizations/graph-optimizations.html#layout-optimizations
+      CHECK_STATUS(ort_api->SetSessionGraphOptimizationLevel(
+          session_options, GraphOptimizationLevel::ORT_ENABLE_BASIC));
+      break;
+    }
+    case mojom::CreateContextOptions::Device::kGpu:
+    case mojom::CreateContextOptions::Device::kNpu: {
+      // It is recommended to disable the graph optimization for OpenVINO backend.
+      // https://onnxruntime.ai/docs/execution-providers/OpenVINO-ExecutionProvider.html#other-configuration-settings
+      CHECK_STATUS(ort_api->SetSessionGraphOptimizationLevel(
+          session_options, GraphOptimizationLevel::ORT_DISABLE_ALL));
+
+      std::string device_type =
+          context_options->device == mojom::CreateContextOptions::Device::kGpu
+              ? "GPU"
+              : "NPU";
+      OrtOpenVINOProviderOptions openvino_options;
+      openvino_options.device_type = device_type.c_str();
+
+      CHECK_STATUS(ort_api->SessionOptionsAppendExecutionProvider_OpenVINO(
```

Comment:
If OpenVINO is not available, this call would fail. I suppose we should return an error instead of crash for such case?

Reply:
Maybe we should try to append the ov ep when creating context with GPU/NPU type, if fail, report error earlier, rather than here when building graph? That can be a TODO